### DISCUSSION TITLE
Is it possible to have Preset caching persist across repositories renovation?

### DISCUSSION THREAD

**Discussion by lyonlai** [renovatebot/renovate#23674] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

renovate/renovate:36.26.1-slim

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Preset resolving persisted across repository renovation in order to reduce number of Github API calls. Perhaps making the config resolving cache uses something similar to the package cache would be good. Thought?

### What you tried so far.

What I have found so far is that.

* there are caches for preset fetching.
* [preset caching is using memcache](https://github.com/renovatebot/renovate/blob/main/lib/config/presets/index.ts#L224C38-L224C38).
* [memcache will be reset between renovating repositories](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/init/cache.ts#L10) 

This results in presets being fetched over and over again between repository in the same Renovate run, even if the content of the preset doesn't change. This causes some stress in Github API Limit for us. And down the road as the config complexity goes up it will get worse. 

I also had a look at the [globalExtends](https://docs.renovatebot.com/self-hosted-configuration/#globalextends) option. It is good that it only resolved once across the whole run. However, my understanding of using that is `ignorePresets` in repository level config wouldn't work against it. Please correct me if my understanding is not correct.

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
> I also had a look at the [globalExtends](https://docs.renovatebot.com/self-hosted-configuration/#globalextends) option. It is good that it only resolved once across the whole run. However, my understanding of using that is `ignorePresets` in repository level config wouldn't work against it. Please correct me if my understanding is not correct.

Yes, globalExtends means the preset is resolved during initialization. A repo can't ignore a preset which is already resolved

**Comment by lyonlai** [comment]
So that wouldn't be ideal. We still prefer to have the ability to ignore the preset. 

What's your thought about making the preset caching not to use memcache but something similar to package cache that would at least be retained for a given renovate run.

**Reply by rarkins** [reply]
Two possibilities:
- Cache it for a static period of time (e.g. 15 minutes)
- Cache it for the entire run
- Cache it, but check if it's changed on each repo using `If-Modified-Since` header to avoid API rate limits

**Reply by viceice** [reply]
I think caching for the run is the simplest way. we can do it at a single place inside preset fetching code.

maybe we should add a second global memory cache and hold run data there. something like we do for host rules

**Reply by rarkins** [reply]
Should it be default behavior?
Should it be configurable?

**Reply by lyonlai** [reply]
I'm thinking maybe the default? It seems to be beneficial in general. e.g. reducing the number of required github API calls and also make the renovate run more stable against the config changes during the run.

Would you prefer some contribution or leave it to you?

**Reply by rarkins** [reply]
Definitely a contribution would be better here. We run the hosted app with one run per job so we have no reason to implement this ourselves.

If the main goal is API rate limits then my third approach above should in theory achieve that while also not resulting in anyone using "stale" presets. I'm not sure where the best abstraction for this would be though, because the "fetch a file" API is implemented separately per platform.

On the other hand a fixed cache period like 15 minutes would achieve almost all the same reduction in API calls while also meaning results can't get _that_ stale. This would also be better for Renovate On-premises which runs as a long-running process and _never_ restarts.

**Reply by lyonlai** [reply]
I didn't notice that Renovate On-Premises could be a long-running process that never restarts. I thought the process would end once it's gone through the whole list of repositories specified in the config.

15 mins cache make sense. Maybe make the length of cache configurable? like for our use case I'll dial it up to an hour.

in terms of the point of the abstraction, maybe just replace the current memcache with some global memory cache and set the item validity according to the configuration? Then it wouldn't go deep into the platform fetch file details.

**Reply by rarkins** [reply]
Yes, if you make it a wrapper of the preset fetching at the worker level then it will avoid changing platform code. If you take a look and have a proposed solution, please let us know so we can give feedback before starting work (unless it's a real little amount of work you don't mind). I want to avoid anyone doing significant work which requires design changes later

**Reply by lyonlai** [reply]
Sure. So I had a look at how preset is fetched and cached. It is pretty straightforward. The following section is how preset's fetched and cached currently using the `memCache` from `../../util/cache/memory`

https://github.com/renovatebot/renovate/blob/main/lib/config/presets/index.ts#L224-L233

Combining with what @viceice suggested we could have something similar to the `hostRules` in memory cache in 

https://github.com/renovatebot/renovate/blob/main/lib/util/host-rules.ts#L10

This type of cache setup should be sufficient to hold the preset data until the Renovate process finishes. This already fits my need.

Configurable cache validity could be added if you think it is important to have @rarkins .

**Reply by viceice** [reply]
I think clearing the global cache after main repo processing loop, like we do for host rules should be enough

**Reply by rarkins** [reply]
We would need either:
- preset caching is configurable on/off, or
- protection against cached private presets (we don't want to resolve them for repos without credentials)

**Reply by lyonlai** [reply]
I'll make the configurable on/off switch for this functionality. 

Protection against cached private presets is important but seems to be some decent amount of work down the road.

**Reply by rarkins** [reply]
Sure, so I guess making it an opt-in function which caches all preset fetches for 15 minutes works best for now.

A simple approach is to keep it in global state, alternatively using the package cache would allow multiple instances to share it. I don't mind which

**Reply by lyonlai** [reply]
package cache seems quite handy with a second look at it.

* clean up after repository processing is done. 
* able to share across different instances which could benefit us down the road if we enabled using `redis` store down the road.

By using the package cache it also enables us the `cacheTtlOverride` option to configure the ttl we need for preset fetching. But it's an experimental feature, would you suggest a brand new option for preset fetching ttl or using this experimental option is ok?

**Reply by rarkins** [reply]
I think you should have a fixed TTL for now

**Reply by viceice** [reply]
15min would be too less, my bot need 30min to run. 😕

**Reply by lyonlai** [reply]
for my case I'll set it to 60mins. maybe make use of the `cacheTtlOverride` for the need of override?
